### ISSUE TITLE
MS Exchange AUTH response can confuse PHPMailer

### ISSUE THREAD

**Issue by greecemunky** [PHPMailer/PHPMailer#462]
On a Microsoft Exchange 2010 server with a receive connector that marks a client IP address as "externally authenticated", after a EHLO the server returns the following.

```
250-my.msexch.server Hello [192.168.123.123]
250-SIZE 1048576000
250-PIPELINING
250-DSN
250-ENHANCEDSTATUSCODES
250-AUTH
250-8BITMIME
250-BINARYMIME
250-CHUNKING
250-XEXCH50
250 XSHADOW
```

Which is stored as in ```$this->server_caps```:

```
Array
(
    [EHLO] => my.msexch.server
    [SIZE] => 1048576000
    [PIPELINING] => 1
    [DSN] => 1
    [ENHANCEDSTATUSCODES] => 1
    [AUTH] => 1
    [8BITMIME] => 1
    [BINARYMIME] => 1
    [CHUNKING] => 1
    [XEXCH50] => 1
    [XSHADOW] => 1
)
```

If (accidentally or incorrectly) configuring PHPMailer to use SMTP Auth from this client, (I assume) because "AUTH" is present but doesn't have any methods listed, PHPMailer errors with the following:

```
PHP Warning:  implode(): Invalid arguments passed in class.smtp.php on line 386
PHP Warning:  in_array() expects parameter 2 to be array, boolean given in class.smtp.php on line 392
PHP Warning:  in_array() expects parameter 2 to be array, boolean given in class.smtp.php on line 392
PHP Warning:  in_array() expects parameter 2 to be array, boolean given in class.smtp.php on line 392
PHP Warning:  in_array() expects parameter 2 to be array, boolean given in class.smtp.php on line 392
```

phpmailerException: ```<strong>SMTP Error: Could not authenticate.</strong><br />```

Disabling SMTP auth from the same client to the same server resolves the issue.

An additional check on ```$this->server_caps['AUTH']``` may be required to confirm it's an array prior to attempting to auth.

**Comment by Synchro** [comment]
That's an RFC2554 contravention (what a surprise, coming from Exchange!) as an empty auth-type is not allowed, but you're right that it should be trapped.

**Comment by greecemunky** [comment]
@Synchro Thanks for the heads-up on Exchange not complying with RFC2554. I wasn't aware.

I've tested a small snippet that can be inserted between the ```if (!array_key_exists('AUTH', $this->server_caps)) { }``` and ```self::edebug('Auth method requested: ' . ($authtype ? $authtype : 'UNKNOWN'), self::DEBUG_LOWLEVEL);``` fragments:

```
if (!is_array($this->server_caps['AUTH']))  {
    $this->setError('Server returned an invalid list of authentication methods');
    return false;
}
```

**Comment by Synchro** [comment]
I've pushed a fix for this. I implemented a bit differently than you suggested - making it deal with the AUTH type fields specially when they are parsed, resulting in an empty array in server_caps if no AUTH types are listed - your Exchange server should now fail with an unknown auth type error. If you find that's not workable we could instead remove the AUTH capability from server_caps if the type list is empty.
